core/node/state_keeper/src/keeper.rs

```diff
@@ -943,3 +843,265 @@ impl ZkSyncStateKeeper {
+                StateKeeperInner::start_next_l2_block(updates_manager, batch_executor).await?;
+            }
+
+            tracing::debug!(
+                "L2 block #{} should be sealed as per L1 batch unconditional sealing rules",
+                updates_manager.l2_block.number,
+            );
+            return Ok(Some(ProcessBlockIterationOutcome::SealBatch));
+        }
+
+        if !updates_manager.has_next_block_params()
+            && inner.io.should_seal_l2_block(updates_manager)
+        {
+            tracing::debug!(
+                "L2 block #{} should be sealed as per L2 block sealing rules",
+                updates_manager.l2_block.number,
+            );
+            return Ok(Some(ProcessBlockIterationOutcome::SealBlock));
+        }
+
+        let waiting_latency = KEEPER_METRICS.waiting_for_tx.start();
+        if let Some(next_l2_block_timestamp) = updates_manager.next_l2_block_timestamp_ms_mut() {
+            // The next block has not started yet, we keep updating the next l2 block parameters with correct timestamp
+            inner
+                .io
+                .update_next_l2_block_timestamp(next_l2_block_timestamp);
+        }
+
+        let Some(tx) = inner
+            .io
+            .wait_for_next_tx(
+                POLL_WAIT_DURATION,
+                updates_manager.get_next_l2_block_or_batch_timestamp(),
+            )
+            .instrument(info_span!("wait_for_next_tx"))
+            .await
+            .context("error waiting for next transaction")?
+        else {
+            waiting_latency.observe();
+            tracing::trace!("No new transactions. Waiting!");
+            return Ok(None);
+        };
+        waiting_latency.observe();
+
+        let tx_hash = tx.hash();
+
+        // We need to start a new block
+        if updates_manager.has_next_block_params() {
+            StateKeeperInner::start_next_l2_block(updates_manager, batch_executor).await?;
+        }
+
+        let (seal_resolution, exec_result) = inner
+            .process_one_tx(batch_executor, updates_manager, tx.clone())
+            .await?;
+
+        let latency = KEEPER_METRICS.match_seal_resolution.start();
+        match &seal_resolution {
+            SealResolution::NoSeal => {
+                let TxExecutionResult::Success {
+                    tx_result,
+                    tx_metrics: tx_execution_metrics,
+                    call_tracer_result,
+                    ..
+                } = exec_result
+                else {
+                    unreachable!(
+                        "Tx inclusion seal resolution must be a result of a successful tx execution",
+                    );
+                };
+                updates_manager.extend_from_executed_transaction(
+                    tx,
+                    *tx_result,
+                    *tx_execution_metrics,
+                    call_tracer_result,
+                );
+            }
+            SealResolution::IncludeAndSeal => {
```

Comment:
Dumb question: Why was this split off from the `SealResolution::NoSeal` match hand above? AFAICT, the code executed is still identical in both cases.